File: 0785-is-graph-bipartite/0785-is-graph-bipartite.cpp
class Solution {
public:

    bool dfs(int curr , int currcolor , vector<int> &color , vector<vector<int>> &adj){
        color[curr] = currcolor;

        for(int &v : adj[curr]){
            if (color[v] == currcolor) return false;
            
            if (color[v] == -1){
                if( !dfs(v , 1-currcolor , color , adj)) return false;
            }

        }

        return true;

    }

    bool isBipartite(vector<vector<int>>& graph) {
        vector<int> color(graph.size() , -1 );
        for(int i = 0 ; i < graph.size() ; i++){
            if(color[i] == -1){
                if(!dfs(i , 0 , color , graph)) return false;
            }
        }

        return true;

    }
};

File: 0785-is-graph-bipartite/0785-is-graph-bipartite.py
class Solution:
    def isBipartite(self, graph: List[List[int]]) -> bool:
        def dfs(curr , currcolor , color , adj):
            color[curr] = currcolor

            for v in adj[curr]:
                if color[v] == currcolor:
                    return False
                
                if (color[v] == -1):
                    if (not dfs(v , 1-currcolor , color , adj)):
                        return False
            
            return True
        

        n = len(graph)
        color = [-1] * n

        for i in range(n):
            if(color[i] == -1):
                if(not dfs(i , 0 , color , graph)):
                    return False
        
        return True